rapthor/lib/cluster.py

```diff
@@ -3,118 +3,178 @@
-        # only Dutch stations
-        mem_usage_gb = 0.83e-3
-    elif antenna == 'LBA':
-        # Memory usage in GB/chan/timeslot/dir of a typical LBA observation
-        mem_usage_gb = 2.5e-4
-    gb_per_solint = mem_usage_gb * solint_slow_freqstep * solint_slow_timestep * ndir
-
-    # Determine if we need to reduce solint_slow_timestep to fit in memory. We adjust
-    # the time step rather than the frequency step, as it is less critical and usually
-    # has a finer sampling
-    if mem_gb / gb_per_solint < 1.0:
-        old_solint_slow_timestep = solint_slow_timestep
-        solint_slow_timestep *= mem_gb / gb_per_solint
-        solint_slow_timestep = max(1, int(round(solint_slow_timestep)))
+    def get_gb_per_solution(target_freqstep, target_timestep, ndir):
+        # Estimate the memory usage in GB per solution
+        #
+        # Note: the numbers below were determined empirically from typical (Dutch-only)
+        # datasets
+        if antenna == 'HBA':
+            # Best-fit coefficients for 25 time slots
+            coef = [7.33333333e-01, 8.33333333e-02, 7.66666667e-03, 5.50000000e-03,
+                    -3.20833714e-17, -4.39443416e-18]
+        elif antenna == 'LBA':
+            # TODO: determine best-fit coefficients
+            # (for now set to HBA values)
+            coef = [7.33333333e-01, 8.33333333e-02, 7.66666667e-03, 5.50000000e-03,
+                    -3.20833714e-17, -4.39443416e-18]
+        gb_per_25_timeslots = (coef[4]*ndir**2 + coef[5]*target_freqstep**2 +
+                               coef[3]*ndir*target_freqstep + coef[1]*ndir +
+                               coef[2]*target_freqstep + coef[0])
+
+        # Return total usage, scaled to the requested time interval, plus a 10% safety
+        # buffer
+        return gb_per_25_timeslots * target_timestep / 25 * 1.1
+
+    # Determine whether we need to adjust the solution interval in time to fit
+    # the solve in memory. We adjust the time interval rather than the frequency
+    # one, as it is less critical and usually has a finer sampling
+    mem_gb = get_available_memory()
+    gb_per_sol = get_gb_per_solution(target_freqstep, target_timestep, ndir)
+    if mem_gb / gb_per_sol < 1.0:
+        solint = target_timestep * mem_gb / gb_per_sol
+        solint = max(1, int(round(solint)))
+    else:
+        solint = target_timestep
+    if solint < target_timestep:
         log.warn('Not enough memory available for slow-gain solve. Reducing solution '
-                 'time interval from {0} to {1}'.format(old_solint_slow_timestep,
-                                                        solint_slow_timestep))
-    nsolints = max(1, int(round(mem_gb / gb_per_solint)))
-    channelsperchunk = np.ceil(solint_slow_freqstep * nsolints)
-    target_freq_chunksize = channelwidth * channelsperchunk
+                 'time interval from {0} to {1} time slots'.format(target_timestep, solint))
```

Comment:
Again, I think the `log.warn()` can be moved safely into the if-clause of line 153.